### PR TITLE
#3665 Functions with dot in name

### LINKED ISSUE #3665: functions with dot in name
Hi there,

you must handle dots in function names. AWS does not allow them.

How to reproduce

> serverless create --template aws-java-maven --path my-lambda-sls
> cd my-lambda-sls/
> sed -i 's/hello:/hello.test:/' serverless.yml 
> mvn clean package
> serverless deploy -v

ends in 

> Serverless: Packaging service...
> Serverless: Uploading CloudFormation file to S3...
> Serverless: Uploading artifacts...
> Serverless: Updating Stack...
>  
>   Serverless Error ---------------------------------------
>  
>      Template format error: Resource name Hello.testLogGroup
>      is non alphanumeric.
>  
>   Get Support --------------------------------------------
>      Docs:          docs.serverless.com
>      Bugs:          github.com/serverless/serverless/issues
>      Forums:        forum.serverless.com
>      Chat:          gitter.im/serverless/serverless
>  
>   Your Environment Information -----------------------------
>      OS:                 linux
>      Node Version:       6.9.1
>      Serverless Version: 1.13.2`

Best regards

### PR DESCRIPTION
**Pull request by RafalWilinski**
## What did you implement:

Closes #3665

Replaces `.` with `Dot` when normalising names and `.` to `-dot-` for AWS function names.

## How did you implement it:
Changed few things inside `/lib/naming.js`, implemented ` getFunctionNameWithoutDots` function.

## How can we verify it:

```
serverless create --template aws-java-maven --path my-lambda-sls
cd my-lambda-sls/
sed -i 's/hello:/hello.test:/' serverless.yml
mvn clean package
serverless deploy -v
```

Without `getFunctionNameWithoutDots` function following error occurred:

```
An error occurred while provisioning your stack: HelloDottestLambdaFunction - 1 validation error detected: Value 'my-lambda-sls-dev-hello.test' at 'functionName' failed to satisfy constraint: Member must satisfy regular expression pattern: (arn:(aws|aws-us-gov):lambda:)?([a-z]{2}(-gov)?-[a-z]+-\d{1}:)?(\d{12}:)?(function:)?([a-zA-Z0-9-_]+)(:(\$LATEST|[a-zA-Z0-9-_]+))?.
```

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

**Comment by eahefnawy** [comment]
@RafalWilinski thanks for that fix!K 🙌 

I see that you've only updated the lambda name in the compile step, which should take care of packaging and deployment. But what about invoking functions with a dot? and logging? and other plugins? shouldn't they be aware of that normalized function name as well?

**Comment by RafalWilinski** [comment]
You are correct, I totally forgot about that, thanks! I'll try to update this PR asap.

File: lib/classes/Service.js [hunk 1]
```diff
@@ -113,7 +113,12 @@ class Service {
         that.custom = serverlessFile.custom;
         that.plugins = serverlessFile.plugins;
         that.resources = serverlessFile.resources;
-        that.functions = serverlessFile.functions || {};
+        that.functions = {};
+
+        // normalize functions with `.` in name to `-dot-`
+        _.forIn(serverlessFile.functions, (value, key) => {
+          that.functions[key.replace(/[.]/g, '-dot-')] = value;
```
Comment (RafalWilinski) [inline lib/classes/Service.js]: I feel this it's not elegant, any ideas where I should move that? Is it even right direction?

**Comment by RafalWilinski** [comment]
Closing for now to keep things tidy. I'll jump on that soon and submit fresh PR

**Comment by pmuens** [comment]
> Closing for now to keep things tidy. I'll jump on that soon and submit fresh PR

Thanks for your work on this one @RafalWilinski :+1: Looking forward to the new PR. Let us know if you need any help with that 👍

**Pull request closed without merging**